torch/csrc/jit/tensorexpr/bounds_inference.cpp

```diff
@@ -0,0 +1,172 @@
+namespace tensorexpr {
+
+class BoundsInference : public IRVisitor {
+ public:
+  void visit(const FunctionCall* v) override;
+  void visit(const Load* v) override;
+  void visit(const Store* v) override;
+  void visit(const For* v) override;
+  void visit(const Block* v) override;
+
+  BoundsInfo accesses() const {
+    return accesses_;
+  }
+
+ private:
+  BoundsInfo accesses_;
+};
+
+void BoundsInference::visit(const Load* v) {
+  accesses_.push_back({v->buf(), kLoad, v->indices(), v->indices()});
+}
+
+void BoundsInference::visit(const FunctionCall* v) {
+  accesses_.push_back(
+      {v->tensor()->func_var(), kLoad, v->params(), v->params()});
+}
+
+void BoundsInference::visit(const Store* v) {
+  accesses_.push_back({v->buf(), kStore, v->indices(), v->indices()});
+  IRVisitor::visit(v);
+}
+
+void BoundsInference::visit(const For* v) {
+  v->body()->accept(this);
+  for (TensorAccessBoundsInfo& access : accesses_) {
+    for (size_t j = 0; j < access.start.size(); j++) {
+      // TODO: This function assumes that all indices grow monotonically and
+      // thus for the loop:
+      //   for i in A..B:
+      //     buf[i] = i
+      // the range for i is [A, B). It should be generalized to correctly handle
+      // all cases.
+      const Expr* old_start = access.start[j];
+      const Expr* old_stop = access.stop[j];
+      const Expr* new_start = Substitute(old_start, {{v->var(), v->start()}});
+      const Expr* new_stop =
+          Substitute(old_stop, {{v->var(), new Sub(v->stop(), new IntImm(1))}});
+      access.start[j] = IRSimplifier::simplify(new_start);
+      access.stop[j] = IRSimplifier::simplify(new_stop);
+    }
+  }
+}
+
+void BoundsInference::visit(const Block* v) {
+  BoundsInfo res;
+  for (auto s : v->stmts()) {
+    s->accept(this);
+    res.insert(res.end(), accesses_.begin(), accesses_.end());
+  }
+  accesses_ = res;
+}
+
+void printBoundsInfo(const BoundsInfo& v) {
+  std::cerr << "Access vector {\n";
+  for (const auto& b : v) {
+    std::cerr << *b.buf << " in (";
+    int i = 0;
+    for (const auto& s : b.start) {
+      if (i != 0) {
+        std::cerr << ", ";
+      }
+      std::cerr << *s;
+      i++;
+    }
+    std::cerr << "; ";
+    i = 0;
+    for (const auto& s : b.stop) {
+      if (i != 0) {
+        std::cerr << ", ";
+      }
+      std::cerr << *s;
+      i++;
+    }
+    std::cerr << ")\n";
+  }
+  std::cerr << "}\n";
+}
+
+// TODO: This probably should be done as a part of IR simplifier.
+static const Expr* simplifyMin(const Expr* a, const Expr* b) {
+  const Expr* diff = IRSimplifier::simplify(new Sub(a, b));
+  if (auto diff_imm = dynamic_cast<const IntImm*>(diff)) {
+    if (diff_imm->value() < 0) {
+      return a;
+    } else {
+      return b;
+    }
+  }
+  return new Min(a, b, true);
+}
+
+static const Expr* simplifyMax(const Expr* a, const Expr* b) {
+  const Expr* diff = IRSimplifier::simplify(new Sub(a, b));
+  if (auto diff_imm = dynamic_cast<const IntImm*>(diff)) {
```

Comment:
This should probably be `isConstant`, although it feels like using the inferred bounds could get us a little closer in some instances.